### PR TITLE
Fix redundant LieGroup gradient dimensions in second-order optimizers

### LINKED ISSUE #387: Reduce the redundant dimension of gradient for LieGroup optimization
## Summary

Currently, when optimizing a LieGroup parameter using `pp.optim`, the gradient's shape is always aligned with the parameter's shape by padding zeros at the end due to the following way of handling parameter updates.

https://github.com/pypose/pypose/blob/33cb63a9c42a6d29aa168e913d3e7cdd597a60ca/pypose/optim/optimizer.py#L96-L101

These zeros are then discarded in the corresponding `ltype.add_` method (e.g. the `SE3Type.add_` below dropped the 7-th col of gradient since they are always zero):

https://github.com/pypose/pypose/blob/33cb63a9c42a6d29aa168e913d3e7cdd597a60ca/pypose/lietensor/lietensor.py#L443-L444

While this design maintains the compatibility with torch's first order optimizers, the padded zeros significantly degrades the numeric condition of second-order optimizers. (Since the zero-padded jacobians $J$ are used to estimate hessian by $H\approx J^TJ$, lead to singular matrix in the solver)

This padded zero can be removed by using the following logic in second-order optimizer base class `_Optimizer`:

```python
    def update_parameter(self, params, step):
        """
        params will be updated by calling this function
        """
        def get_gradient_size(p: torch.Tensor | pp.LieTensor) -> torch.Size:
            if isinstance(p, pp.LieTensor):
                return p.shape[:-1] + p.ltype.manifold
            else:
                return p.shape
        
        grad_params = [p for p in params if p.requires_grad]
        grad_shapes = [get_gradient_size(p) for p in grad_params]
        steps       = torch.split(step, [s.numel() for s in grad_shapes])
        for p, d, s in zip(grad_params, grad_shapes, steps):
            p.add_(s.view(d))

```

## Improvements

Experiment shows that removing the padding zeros in the jacobian significantly improve the numeric stability of estimated hessian matrix

Below is the numeric stability analysis for backend optimizer of a visual odometry for the `A` matrix on this line:

https://github.com/pypose/pypose/blob/33cb63a9c42a6d29aa168e913d3e7cdd597a60ca/pypose/optim/optimizer.py#L284

Current implementation:

Image

Removing the redundant zero cols:

Image

The condition number improved significantly (average condition number from the sampling period reduced from $1\times 10^{14}$ to $1\times 10^{5}$). Meaning that the solution is significantly more stable.

## Risks

(Please list features that can break because of this refactoring and how you intend to solve that.)

## Involved components

(List files or directories that will be changed by the refactoring.)

## Optional: Intended side effects

(If the refactoring involves changes apart from the main improvements, list them here.
It may be a good idea to create separate issues and link them here.)

## Optional: Missing test coverage

(
If you are aware of tests that need to be written or adjusted apart from unit tests for the changed components,
please list them here.
)

### PR DESCRIPTION
**Pull request by alphapibeta**
## Summary

Fix redundant LieGroup embedding dimensions in second-order optimizer Jacobians and parameter updates.

For LieTensor parameters such as SE3, the embedded representation contains redundant coordinates (7 values) while the tangent/manifold update has 6 degrees of freedom. The previous implementation retained the padded embedding columns, making the estimated Hessian unnecessarily singular or poorly conditioned.

## Changes

- Slice redundant LieGroup embedding coordinates before flattening Jacobian rows.
- Use `p.shape[:-1] + p.ltype.manifold` for LieTensor optimizer step splitting and reshaping.
- Preserve ordinary Tensor behavior.
- Correct dense and sparse LM parameter-step alignment.
- Exclude frozen parameters consistently from Jacobian columns and step updates.
- Add regression coverage for batched SE3, ordinary Tensor, mixed parameters, and frozen parameters.

## Validation

CPU-only validation with Torch 2.13.0+cu130 and PyPose editable-installed in the dedicated development environment:

- `tests/optim/test_optimizer.py`: 17 passed.
- Full suite: **97 passed, 2 skipped**.
- `git diff --check`: passed.

The two skipped tests are existing expected skips. The test run used `CUDA_VISIBLE_DEVICES=""`; no GPU was used.

Fixes #387

### CODE REVIEW

**Comment by wang-chen** [comment]
@zitongzhan this could be a potential very useful fix, could you check whether it fits BAE as well?

**Comment by zitongzhan** [comment]
> @zitongzhan this could be a potential very useful fix, could you check whether it fits BAE as well?

No. `flatten_row_jacobian` was not used by `bae` because it is for dense Jacobian and cannot be used for BA.

**Review by zitongzhan — CHANGES_REQUESTED** [review]
talked with Chen. i am pretty sure the PR is correct because bae had that design pattern 2 yrs ago. Apart from correctness, i am uncertain about compliances of coding style which Chen will enforce. Those are primarily simplicity nit picks.

File: pypose/optim/optimizer.py [hunk 1]
```diff
@@ -74,9 +74,19 @@ def __init__(self, model, kernel=None, auto=False):
 
     def flatten_row_jacobian(self, J, params_values):
         if isinstance(J, (tuple, list)):
-            J = torch.cat([j.reshape(-1, p.numel()) for j, p in zip(J, params_values)], 1)
+            # Only include trainable parameters to match update_parameter split.
+            pairs = [(j, p) for j, p in zip(J, params_values) if p.requires_grad]
+            J = torch.cat([self._flatten_single_jacobian(j, p) for j, p in pairs], 1)
         return J
```
[the remark this answers is not in the document]
Comment (alphapibeta) [reply pypose/optim/optimizer.py:84]: Addressed in 5e6924a. flatten_row_jacobian now raises a clear RuntimeError when no trainable parameters remain, instead of reaching torch.cat([]). The focused optimizer tests and full test suite pass.

File: .readthedocs.yaml [hunk 2]
```diff
@@ -7,7 +7,7 @@ version: 2
 
 # Set the version of Python and other tools you might need
 build:
-  os: ubuntu-20.04  # <- add this line
+  os: ubuntu-22.04
   tools:
     python: "3.9"
```
[the remark this answers is not in the document]
Comment (alphapibeta) [reply .readthedocs.yaml:12]: Read the Docs configuration was updated from Ubuntu 20.04 to Ubuntu 22.04 because the current RTD validator rejects Ubuntu 20.04; this does not affect runtime behavior.

File: tests/optim/test_optimizer.py [hunk 3]
```diff
@@ -359,6 +359,130 @@ def forward(self, poses):
...
+    def test_frozen_parameters_excluded(self):
+        """Parameters with requires_grad=False are excluded from updates."""
+        class PoseInv(nn.Module):
+            def __init__(self, *dim):
+                super().__init__()
+                self.pose1 = pp.Parameter(pp.randn_se3(*dim).Exp())
+                self.pose2 = pp.Parameter(pp.randn_se3(*dim).Exp())
+
+            def forward(self, inputs):
+                return (self.pose1 @ inputs).Log()
```
[the remark this answers is not in the document]
Comment (alphapibeta) [reply tests/optim/test_optimizer.py]: The suggested patch no longer applies because the code was already changed in ef2883c

**Review by zitongzhan — APPROVED** [review]
I think the PR is concise because it reuses the recent _parameter_update_shape introduced in bae PR. I have no further doubts
